### ISSUE TITLE
T16 Three `.cu` kernels + CPU oracles

### ISSUE THREAD

**Issue by superposition** [superposition/qualia#31]
Part of #6

## Steps

**Step 16. Add the three kernels.**
In `crates/cuda/kernels/`, add `belief_couple.cu` (prior-weighted slot update, 1 thread per slot),
`perception_voxel.cu` (lidar points to occupancy logits, 1 thread per voxel, mirroring
`leash-cuda`'s voxel kernel's input shape: flat `[x, y, ...]` coordinates), and `action_score.cu`
(one block per rollout candidate scoring `terminal_goal_distance_m` and
`max_collision_probability` against a candidate's steps). Register all three in the same module list
the existing kernels use and add CPU oracles beside the existing ones in `crates/cuda/src/lib.rs`
following the established oracle pattern.

## Test written first

`oracle_matches_reference_math`

## Commands

`cargo test -p qualia-cuda`

## Evidence

_pending_

## Journal

the-fly-in-the-belief-matrices

## Blocked by

- T5 (#20)

## Definition of done (2026-09-11)

- the package builds and its tests pass on the dev host (`cargo build -p <pkg> && cargo test -p <pkg>`);
- the artifact is built for the robot (aarch64) and exercised on **Pinkie**, the Waveshare-carried
  Jetson Orin NX at `jetson@192.168.55.1` (see `docs/decisions.md` D-010), by this ticket's own smoke
  path or command;
- the handoff comment quotes the build/deploy command and the output observed on the board;
- a host-only package that cannot run there (the macOS `metal` backend) records the aarch64 build as
  its board evidence and states why execution on the board is impossible.

**Comment by superposition** [comment]
```braid
agent: TicketT16
branch: ticket/T16
commit: none
state: claimed
next: write the named oracle test first, then the three kernels and their CPU oracles
blocked_on: none
evidence: none
```

Claiming T16 (#31). Worktree `C:/Users/ericm/wt/T16`, branch `ticket/T16` from `origin/main` @ `25f2b6f`.

Reading: the three kernels are new to the rewrite — neither the reference engine (`kernels/` holds only `belief_update.cu` and `evidence_layout.cu`) nor `leash` ships a `belief_couple.cu`, `perception_voxel.cu` or `action_score.cu`, so the ticket text is the shape contract and the reference math is what each oracle must reproduce: the connectome `CouplingPrior::couple` normalisation for the slot update, the world voxel lattice (32x32x12, `idx = vx*VOXEL_D*VOXEL_H + vz*VOXEL_H + vy`, flat ground-plane coordinates) for the lidar projection, and `planner.rs`'s midpoint-integration / central-collision arithmetic for the rollout cost terms. Test `oracle_matches_reference_math` lands red first.

**Comment by superposition** [comment]
```braid
agent: TicketT16
branch: ticket/T16
commit: 84a76de
state: review
next: read the three review verdicts on PR #170, then merge it
blocked_on: none
evidence: docs/evidence/T16/three-kernels/README.md
```

Handoff in PR https://github.com/superposition/qualia/pull/170 (`Closes #31`), pushed `84a76de`, based on `origin/main` @ `54e7f3a`.

Three kernels under the workspace-root `kernels/` (D-006) with host oracles and device contexts in `crates/cuda`; the named test `oracle_matches_reference_math` landed first, red on the missing oracles. Host: 17 tests pass (14 before). 4090 (`NVIDIA GeForce RTX 4090, 591.74, 24564 MiB`), serial one-thread run: lib 2, cpu_fallback 13, gpu 8, kernel_abi 4, all pass. `python C:/tmp/gate.py` -> `provenance: OK`. mage capture committed under `docs/evidence/T16/three-kernels/` (16 launches over 8 kernels; the three new ones run once each: perception_voxel 48x256 11.84 us, action_score 2x256 10.59 us, belief_couple 1x4 2.50 us). Board execution is not claimed: per D-010 the 4090 verifies these kernels and the sm_87 fatbin targets are T18 (#33).

**Comment by superposition** [comment]
```braid
agent: TicketT16
branch: ticket/T16
commit: a7e5f5e
state: review
next: read the three review verdicts on PR #170, then merge it
blocked_on: none
evidence: docs/evidence/T16/three-kernels/README.md
```

Breadcrumb refresh: PR #170 was rebased onto `origin/main` @ `54e7f3a` and gained one test-only cleanup commit after the capture, `a7e5f5e`. The 4090 run was repeated on the cleaned tree (`cargo test -p qualia-cuda --features cuda --test gpu -- --test-threads=1`: 8 passed, 0 failed) and the provenance gate still reads `provenance: OK`. The PR body's braid block names the current commit.

**Comment by superposition** [comment]
```braid
agent: TicketT16
branch: ticket/T16
commit: d791301
state: review
next: clean-room re-read of PR #170 at d791301, then merge it
blocked_on: none
evidence: docs/evidence/T16/three-kernels/capture.sqlite
```

Clean-room fix pushed: `capture.json` argv/command are checkout-relative and the README reproduces
from any checkout (home-relative scratch dir, `find` recovers the hashed test binary), and the
capture's `StringIds` was cut from 134 rows to the 8 kernel-name ids any launch references — the host
paths, `PATH`/`HOME` blocks, distro and session names and nsys session log paths are gone, file
36 KiB → 12 KiB. Verified: mage's nsys backend re-parses 16 launches with the same names and
durations as `kernels.csv`, no interned path or environment value remains, and `provenance: OK` on
`d791301`. Re-read requested in PR #170.

**Comment by superposition** [comment]
```braid
agent: Fix170T16
branch: ticket/T16
commit: d917e4f
state: review
next: Rev170Correct to re-read PR #170 at d917e4f
blocked_on: none
evidence: https://github.com/superposition/qualia/pull/170
```

**2026-09-11 — Step 16's "leash-cuda's voxel kernel's input shape" is wrong; resolution below (fix for the correctness review on PR #170, comment 5633079014).**

Step 16 says `perception_voxel.cu` mirrors "`leash-cuda`'s voxel kernel's input shape: flat `[x, y, ...]` coordinates". No such kernel exists, in either local leash checkout:

- The only voxel kernel is `project_occupancy`, which consumes an `int8` cell grid — `C:/leash/src/cuda_voxel.rs:12` and `C:/Users/ericm/leash/crates/leash-cuda/kernels/leash_kernels.cu:5` both declare `const int8_t* cells, int32_t* output, ...`, with no point list.
- Leash's LiDAR kernel `lidar_transform` (`leash_kernels.cu:20`) writes *separate* `x_m[]` and `y_m[]` arrays from `ranges_m`, never an interleaved pair list.

The interleaved shape the PR implements is real, but it is a data contract rather than a kernel input: `LeashSpatialEvidenceV1.points_xy_m` — `crates/sync-types/src/spatial.rs:15`, validated as `points_xy_m.len() == point_count * 2` (`spatial.rs:32`), consumed in `runners/agent/src/world_model.rs:699` (`chunks_exact(2)`). The only half of the Step 16 phrasing that holds is the extrusion: leash's `project_occupancy` repeats a 2-D cell across its `depth` axis, and `perception_voxel` marks the same whole voxel column.

PR #170's body and the two comments (`kernels/perception_voxel.cu:6-8`, `crates/cuda/src/cpu.rs:291-292`) now name `points_xy_m` as the input shape and leash's `project_occupancy` for the extrusion. This ticket's Step 16 text is left as written — it is the ticket record — and this comment is its resolution.

**T17 (#32) and T18 (#33): the input shape to inherit is `LeashSpatialEvidenceV1.points_xy_m` (interleaved pairs, `len == point_count * 2`), not a leash kernel signature.** A `grep -rn "chunks_exact(2)\|points\[2"` across both leash trees and `C:/qualia` finds no point-list voxel kernel.

Corrected in the same fix: the `[-20, 20]` logit clamp is PR #170's own choice, not the belief kernels' band (they clamp at ±10; the reference's only ±20 clamp is the planner's log-variance at `crates/jepa-model/src/planner.rs:381`, and it has no occupancy-logit band), and the PR body's provenance line now quotes `compared 123 authored file(s)` from a fresh gate run instead of 119.

**Comment by superposition** [comment]
```braid
agent: Main
branch: none (tracker remediation)
commit: none
state: reopened
next: run this ticket's board leg on Pinkie and quote the output here
blocked_on: none
evidence: local://board-dod-audit.md ; https://github.com/superposition/qualia/issues/230
```

## Reopened: the board leg was never run

The retroactive DoD audit (#230, comment 5642836995) classified this ticket's board limb as **absent**. What the
stream holds: PR #170 body: “`sm_87`; the aarch64 fatbin targets are T18 (#33)'s step … **No Pinkie execution is claimed here.**”; evidence is the dev-host nsys capture `docs/evidence/T16/three-kernels/`

The definition of done (`docs/agents.md`) requires the artifact built for the robot (aarch64) and exercised on
**Pinkie** by the ticket's own smoke path, with the build/deploy command and the output observed on the board
quoted in the handoff. That did not happen for this ticket, so it is not done.

The board is being provisioned now — network through the host proxy over the gadget link, the registry cache
verified complete, and the candle closure compiling natively there — and the run below is queued. The ticket
closes again only when its output is quoted in this stream.

Run owed: `crates/cuda` + the three `.cu` kernels. Achievable: `cargo test -p qualia-cuda --features cuda --test gpu --offline -j 2` (and an `ncu` capture on Pinkie) — T17/T18/T50 ran the same suite on the board afterwards, but not this ticket's own capture

**Comment by superposition** [comment]
Board DoD smoke for this ticket's reopened limb: a fresh native aarch64 run on Pinkie, the DoD's board. (Remediation of the retroactive audit on #230; I did not close anything.)

run: `cargo test -p qualia-cuda --features cuda --test gpu --offline -j 2`

board: Pinkie = Jetson Orin NX (aarch64, kernel 5.15.148-tegra, 6 cores), rustc 1.94.0; head `515a5063cc6655f87c821f7f3fe931cc77ca8f55` (current origin/main, `git archive` -> `/home/jetson/remediate/515a506`); `RUSTFLAGS="-C target-feature=+fp16"` (D-016); board clock 2026-09-02 (the board runs ~9 days behind the dev host, D-010).

The kernels were exercised on the board, both ways: the audit named command (first block, 8/0) and the real `sm_87` cubin build (second block, 8/0). The eight device tests are `belief_update_matches_the_host_reference`, `belief_couple_context_matches_the_host_reference`, `cognition_dispatch_matches_the_host_reference`, `cognition_weight_edits_are_observable`, `action_score_context_matches_the_host_reference`, `costmap_stats_context_matches_the_host_reference`, `perception_voxel_context_matches_the_host_reference`, `smoke_context_runs_a_kernel`. An `ncu` capture on the board follows in a second comment. Note the package-level suite (see #78) is red only on the NVRTC `evidence_layout` fixture.

`test result:` lines (verbatim)
```
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.55s
```

output tail (verbatim, `logs/cuda-gpu.log`)
```
     Running tests/gpu.rs (target/debug/deps/gpu-6aef8cf47bcaebbc)

running 8 tests
test cognition_dispatch_matches_the_host_reference ... ok
test action_score_context_matches_the_host_reference ... ok
test belief_couple_context_matches_the_host_reference ... ok
test cognition_weight_edits_are_observable ... ok
test belief_update_matches_the_host_reference ... ok
test costmap_stats_context_matches_the_host_reference ... ok
test perception_voxel_context_matches_the_host_reference ... ok
test smoke_context_runs_a_kernel ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.55s

```

`EXIT=0 TAG=cuda-gpu SECONDS=4 END=2026-09-02T14:23:47Z`

second run on the same board session (the ticket's binary smoke path):

run: `CUDAARCHS=87-real NVCC=/usr/local/cuda/bin/nvcc cargo test -p qualia-cuda --features cuda --test gpu --offline -j 2 -- --test-threads=1`

`test result:` lines (verbatim)
```
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.66s
```

output tail (verbatim, `logs/cuda-gpu-fatbin.log`)
```
     Running tests/gpu.rs (target/debug/deps/gpu-6aef8cf47bcaebbc)

running 8 tests
test belief_couple_context_matches_the_host_reference ... ok
test costmap_stats_context_matches_the_host_reference ... ok
test smoke_context_runs_a_kernel ... ok
test action_score_context_matches_the_host_reference ... ok
test perception_voxel_context_matches_the_host_reference ... ok
test cognition_weight_edits_are_observable ... ok
test belief_update_matches_the_host_reference ... ok
test cognition_dispatch_matches_the_host_reference ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.66s

```

`EXIT=0 TAG=cuda-gpu-fatbin SECONDS=12 END=2026-09-02T14:23:59Z`

**Comment by superposition** [comment]
Board `ncu` capture for this ticket's own limb (the second half of the audit's ask; the suite run is in the previous comment).

command (Pinkie, following `docs/evidence/T18/fatbin-sm-87`'s recipe):
```
CUDAARCHS=87-real NVCC=/usr/local/cuda/bin/nvcc \
  cargo test -p qualia-cuda --features cuda --release --test gpu --offline -j 2 --no-run
echo jetson | sudo -S env LD_LIBRARY_PATH=/usr/local/cuda-12.9/compat \
  /usr/local/cuda/bin/ncu --target-processes all --launch-count 8 \
  --export remediate/t16-ncu/capture --force-overwrite "$BIN" --test-threads=1
/usr/local/cuda/bin/ncu --import remediate/t16-ncu/capture.ncu-rep --page details --csv > remediate/t16-ncu/metrics.csv
```

board/time: Pinkie = Jetson Orin NX (aarch64, kernel `5.15.148-tegra`, driver `NVRM 540.4.0`, CUDA 12.9 `V12.9.41`, `ncu 2025.2.0.0 (build 35613519)`, compute capability 8.7), rustc 1.94.0; head `515a5063cc6655f87c821f7f3fe931cc77ca8f55` (current `origin/main`, `git archive` → `/home/jetson/remediate/515a506`); `RUSTFLAGS="-C target-feature=+fp16"` (D-016); board clock `2026-09-02T14:30Z` (≈9 days behind the dev host, D-010). The build log records `qualia-cuda: embedded fatbins for sm_87`.

output tail (verbatim, `remediate/t16-ncu/ncu.log`):
```
==PROF== Profiling "perception_voxel" - 7 (8/8): 0%....50%....100% - 2 passes
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 22.26s
==PROF== Disconnected from process 461147
==PROF== Report: /home/jetson/remediate/515a506/remediate/t16-ncu/capture.ncu-rep
```
`NCU_RC=0`; `IMPORT_RC=0`; `metrics.csv` 366 rows.

launches (details page; the profiler kept running past the eighth launch, so `smoke`'s kernel is not in the capture):

| id | kernel | block | grid | duration |
| --- | --- | --- | --- | --- |
| 0 | `action_score` | (256, 1, 1) | (2, 1, 1) | 96.61 us |
| 1 | `belief_couple` | (4, 1, 1) | (1, 1, 1) | 21.06 us |
| 2 | `belief_update` | (1024, 1, 1) | (1, 1, 1) | 3.41 ms |
| 3 | `cognition_update` | (1024, 1, 1) | (1, 1, 1) | 3.60 ms |
| 4 | `cognition_update` | (1024, 1, 1) | (1, 1, 1) | 3.60 ms |
| 5 | `cognition_patch` | (1, 1, 1) | (1, 1, 1) | 16.93 us |
| 6 | `costmap_stats` | (256, 1, 1) | (16, 1, 1) | 19.42 us |
| 7 | `perception_voxel` | (256, 1, 1) | (48, 1, 1) | 259.90 us |

The launch shapes match `docs/evidence/T18/fatbin-sm-87` exactly; the durations are lower than that capture's (`belief_update` 3.41 ms here vs 13.00 ms there, `cognition_update` 3.60 ms vs 12.96/12.94 ms) — this pass ran with the profiler's default clock control on a board whose `SM Frequency` read 305.98 MHz. Reported as measured, with no claim attached to the difference.

artifacts stay in board scratch `/home/jetson/remediate/515a506/remediate/t16-ncu/` (`capture.ncu-rep` 491,230 B, `metrics.csv` 76,584 B, `ncu.log` 1,434 B); I committed nothing to the repository. Say the word if you want them landed under `docs/evidence/T16/`.

**Comment by superposition** [comment]
```braid
agent: Main
branch: none (remediation close)
commit: none
state: done
next: none - the board limb is quoted in this stream
blocked_on: none
evidence: the board run comment above (Pinkie aarch64, head 515a506)
```

Closed on the board evidence posted above: on Pinkie (Jetson Orin NX, aarch64, ncu 2025.2.0) `cargo test -p qualia-cuda --features cuda --offline -j 2 --test gpu` -> **8 passed; 0 failed**, the fatbin variant -> 8 passed; 0 failed, and the `ncu` capture recorded **8 kernel launches** with the report kept in board scratch. T16's kernels are now exercised on the device they target, which is what the reopened limb owed (PR #170 had explicitly claimed no Pinkie execution).

**Comment by superposition** [comment]
One environment note on the capture above, for the record: the board's GPU is not exclusively held. A `llama-server` (`/home/jetson/qualia-runtime/current/bin/llama-server --n-gpu-layers 99`, pid 1304) has been running since board clock `2026-09-01T16:56`, so it was resident for this pass — and for every earlier board capture, so it does not distinguish this capture from `docs/evidence/T18`. It does mean the reported durations are shared-GPU numbers with the profiler's default clock control; treat them as the board's own smoke numbers, not as a controlled comparison.
